### ISSUE TITLE
Issue with Poetry virtualenv activation

### ISSUE THREAD

**Issue by mrtolkien** [jdx/rtx#489]
**Describe the bug**
`rtx-poetry` does not properly activate the poetry virtual environment.

**To Reproduce**

`.rtx.toml`:
```toml
[tools]
poetry = { version = '1.4.2', pyproject = 'pyproject.toml' }
python = '3.9'
```

Shell output:

```sh
rtx-poetry: No virtualenv exists. Executing `poetry install` to create one.

[Errno 2] No such file or directory: 'python'
[WARN] Error running exec-env: command ["bash", "-c", ". /Users/tolki/.local/share/rtx/plugins/poetry/bin/exec-env\nexport -p\n"] exited with code 1
/Users/tolki/.local/share/rtx/installs/poetry/1.4.2/bin/poetry env info --path
```

**Expected behavior**

The environment gets loaded properly. `poetry env info` works and the virtual environment is there:
```sh
➜ poetry env info       

Virtualenv
Python:         3.9.16
Implementation: CPython
Path:           /Users/tolki/Library/Caches/pypoetry/virtualenvs/kiminoban-api-BUlJ71KV-py3.9
Executable:     /Users/tolki/Library/Caches/pypoetry/virtualenvs/kiminoban-api-BUlJ71KV-py3.9/bin/python
Valid:          True

System
Platform:   darwin
OS:         posix
Python:     3.9.16
Path:       /Users/tolki/.local/share/rtx/installs/python/3.9.16
Executable: /Users/tolki/.local/share/rtx/installs/python/3.9.16/bin/python3.9
```

**`rtx doctor` output**

```
rtx version:
  1.28.5 macos-arm64 (d206899 2023-04-13)

build:
  Target: aarch64-apple-darwin
  Features: SELF_UPDATE
  Built: Thu, 13 Apr 2023 13:40:55 +0000
  Rust Version: rustc 1.68.2 (9eb3afe9e 2023-03-27)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin22.0)

rtx environment variables:
  RTX_SHELL=zsh

settings:
  {"experimental": "false", "missing_runtime_behavior": "warn", "always_keep_download": "false", "legacy_version_file": "true", "plugin_autoupdate_last_check_duration": "10080", "trusted_config_paths": "[]", "verbose": "false", "asdf_compat": "false", "jobs": "4", "disable_default_shorthands": "false", "log_level": "INFO", "raw": "false"}

config files:
  /Users/tolki/.tool-versions
  /Users/tolki/Development/tachikoma/kiminoban-api/.rtx.toml

plugins:
  poetry       https://github.com/rtx-plugins/rtx-poetry.git#6e1231e
  pre-commit   https://github.com/jonathanmorley/asdf-pre-commit.git#26bfc42
  pulumi       https://github.com/canha/asdf-pulumi.git#99d675b
  python       https://github.com/rtx-plugins/rtx-python.git#2b29b16

toolset:
  poetry@1.4.2, python@3.9

No problems found
```

**Additional context**
I tried fixing the problem by debugging the `rtx-poetry` plugin here:
https://github.com/rtx-plugins/rtx-poetry/blob/main/bin/utils.sh

But `"$(poetry_bin)" env info --path` does return the right value despite `exec-env` executing `rtx-poetry: No virtualenv exists. (...)`.

I am very much unsure as to why the `rtx-poetry` is not picking up the virtual environment properly.

**Comment by jdxcode** [comment]
it looks like the problem is that `python` isn't found because it's only available via rtx and all of the plugins are loaded in parallel.

I think the only good solution here is #393 which would allow poetry to take on python as a dependency. Unfortunately, that's a pretty big challenge.

**Comment by mrtolkien** [comment]
So the workaround is to also have `python` installed as a system-wide dependency without `rtx`?

Would the rtx-managed `poetry` still see all python versions installed by `rtx` in that case?

Right now I use `rtx` to manage my python versions but I use `nix-env` for `poetry` and it's working decently fine despite some weird warning messages:
![image](https://user-images.githubusercontent.com/8489650/232949897-4f519d69-7ba5-4408-994a-3cd66446aa7e.png)

This is with `poetry config virtualenvs.prefer-active-python` being `true`. It still "sees" the nix-installed python 3.10 first despite `rtx` setting the `python` path to the proper value for the folder.

What would be the recommended flow for using `poetry` with `python` in `rtx`?

**Comment by mrtolkien** [comment]
After installing `python3` with `nix-env -iA nixpkgs.python39` it all works without warnings.

*But* it means that:
- The globally (non-rtx) installed `python3` version must match the one required by poetry
- The local python defined by `rtx` has no impact on `poetry` and is not used

It's just an all-around weird situation and I'm not sure what's the "right" way to get Python + poetry working with `rtx`.

Everything works fine with `asdf` with the caveat of being extremely slow and not automatically activating the `poetry` environment so I'd much rather be able to use `rtx` instead!

**Comment by delicb** [comment]
I generally don't use poetry, but I maintain one project that does use it. Instead of using poetry plugin, I just use `rtx-python` plugin to create virtualenv and activate it automatically using `rtx` installed python and have `virtualenvs.create = false` in Poetry config. This way, when I `cd` into a folder, virtualenv is activated and poetry is using that venv, not trying to create a new one. This aligns well with other non-poetry project I have. I understand the benefits of having `poetry` manage its own envs, so this might not be applicable for everyone, just dropping it in case it is useful to someone.

**Comment by jdxcode** [comment]
yeah I think I'd probably go with something like @delicb has for now.

The reason that this isn't a problem for asdf is because it loads everything in series, but in rtx we load everything in parallel so poetry and python are loaded at the same time. We could fix this by removing the parallelization but that would cause a huge hit to performance. The better fix is the ability to declare dependencies so that poetry can say that it needs a python runtime.

**Comment by mrtolkien** [comment]
So that works pretty well but I still get an error message from `poetry` when opening the folder in the VSCode Terminal:

```sh
rtx-poetry: No virtualenv exists. Executing `poetry install` to create one.

[Errno 2] No such file or directory: 'python'
[WARN] Error running exec-env: command ["bash", "-c", ". /Users/tolki/.local/share/rtx/plugins/poetry/bin/exec-env\nexport -p\n"] exited with code 1
```

Here are some other commands to show that everything else is managed properly:
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/8489650/233241557-eacf62ac-ef05-4b84-b71e-5df07a551889.png">

What's the right way to silence that warning?

**Comment by delicb** [comment]
@mrtolkien I am not using `rtx-poetry` plugin, but I see that this message and error is coming from `exec-env` file in `rtx-poetry`. I assume you still have `poetry` section in `.rtx.toml`? In my setup, I have `python` section (essentially, something like this: `python = {version="3.11", virtualenv=".venv"}`) instead of `poetry`. 

This way, poetry just uses already activated virtualenv that rtx activates.

**Comment by mrtolkien** [comment]
I see. What do you use to manage poetry then?

It indeed seems that any way to manage poetry through `rtx` will be faulty at the moment since:
- It always tries to activate a poetry virtual environment
- It can't activate a virtual env with rtx's python version

I really wish everything could be managed with `rtx` because `poetry` made multiple compatibility-breaking changes in the past and I'd like to be able to pin poetry's version too.

**Comment by delicb** [comment]
I agree that it would be good to manage everything via `rtx`, but for now, I just used `pipx` to install poetry (I am using `pipx` anyway for some other tools).

**Comment by RafalSkolasinski** [comment]
Controlling both `python` and `poetry` version per-project is one of main uses of `asdf-vm` for me so this is unfortunately gonna be a blocker for me to migrate. 

> The better fix is the ability to declare dependencies so that poetry can say that it needs a python runtime.

This would sound ideal.

**Comment by mrtolkien** [comment]
Same here, poor packaging tools integration is a bit of a non-starter for using `rtx` for python projects.

As it is structured now any python packaging tool cannot be managed at the same time as python itself with rtx since it searches for python in the path "before" rtx does its job setting it.

**Comment by bryan-lott** [comment]
I'm not entirely following the issue, but we have several python projects using poetry as the dep manager. I've settled on this for our `.rtx.toml` file and haven't seen any of the errors above:

```toml
[tools]
python = { version = "3.9.7", poetry = "pyproject.toml", virtualenv = ".venv" }
```

**Comment by jdxcode** [comment]
@bryan-lott that functionality was removed from rtx-python. However I'm glad you brought that up since an easier solution may be to simply add support for poetry directly into the python plugin like we were doing before.

**Comment by jdxcode** [comment]
I'm not sure though. I like the idea of being able to separately manage versions.

**Comment by bryan-lott** [comment]
@jdxcode Interesting, I keep rtx up to date and I haven't seen any change in functionality (i.e., everything still seems to work okay). Happy to provide other configs and details about my setup if you'd like.

**Comment by jdxcode** [comment]
Your Python plugin is out of date

**Comment by bklebe** [comment]
I'm running into this because I'm using RTX as a drop-in replacement for asdf and we manage Poetry and Python versions both with `.tool-versions` currently. asdf has problems with Poetry and Python as well but because it's not trying to do things in parallel, it works once you `asdf install`. I'd be happy to help fix this if you're looking for any help @jdxcode, I love the project so far and am very impressed with the performance.

**Comment by jdxcode** [comment]
For now I think maybe the best course would be for one if you to fork rtx-python and revert the change I made to remove poetry.

I don't want to do that in the main plugin since I don't want to support that behavior long-term. Dependencies are probably the solution here.

**Comment by khimaros** [comment]
would you mind sharing the changelist that we ought to roll back?

is that this one? https://github.com/rtx-plugins/rtx-python/pull/6

**Comment by jdxcode** [comment]
I think it's this one https://github.com/rtx-plugins/rtx-python/commit/6f71bc5922dd8c42c9808477e5392a06ad6fb6e3

**Comment by jdxcode** [comment]
I had an idea for a slightly simpler version of "dependencies". I realized what we want here isn't so much the ability for a plugin to declare a dependency, it's more that you want poetry to be able to use an existing python iff it is also in .rtx.toml—otherwise it should just use python on PATH.

I think it's similar to peer dependencies in npm so I'm calling it "peers" for now, but I don't love the name. I think this will be a lot simpler than a full-fledged dependency implementation (which would allow specifying version ranges, subdependencies, etc).

Mostly just an idea for now but hopefully I'll find some time to work on it more.

**Comment by barraponto** [comment]
Here is my take on installing poetry with rtx-python: use [RTX_PYTHON_DEFAULT_PACKAGES_FILE](https://github.com/jdx/rtx/blob/main/docs/python.md#default-python-packages). I created a file `.default-python-packages` and added a single line with `poetry` in it. Then I ran:

```sh
RTX_PYTHON_DEFAULT_PACKAGES_FILE=".default-python-packages" rtx install python
```

After that, `which poetry` returns `~/.local/share/rtx/installs/python/3.11.6/bin/poetry`

**Comment by jdx** [comment]
reopened since my solution caused a deadlock and needed to be reverted
